### PR TITLE
feat(event_source): add CloudFormationCustomResourceEvent data class.

### LINKED ISSUE #4332: Feature request: Add CloudFormation CustomResource events to EventSource dataclasses
### Use case

I am building a Lambda handler for managing CDK CustomResources (https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources-readme.html).

### Solution/User Experience

As a user of AWS Lambda Powertools I would like to have dataclass parsing and type-safety for all standard Lambda events coming from an AWS service. CloudFormation custom resources are missing.

### Alternative solutions

_No response_

### Acknowledgment

- [X] This feature request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [X] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by phipag**
**Issue number: #4332**

## Summary
This adds the `CloudFormationCustomResourceEvent` event source data class. There was already a parser but the event source data class was missing.

### Changes

- Add the  `CloudFormationCustomResourceEvent` data class
- Add missing `PhysicalResourceId` to parser

### User experience

As a user I was unable to add a strongly typed event to my Lambda handler for custom CloudFormation resource management.

After the change, I am able to have a typed event for this scenario.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change? No</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]: These values are static and only used once in the class, I think it makes more sense to remove this `Enum` and add it as a `Literal` in the `request_type` property.

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
    def physical_resource_id(self) -> str:
```

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
        return self.get("PhysicalResourceId") or ""
```

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
    def resource_properties(self) -> Dict[str, Any]:
```

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
        return self.get("ResourceProperties") or {}
```

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
    def old_resource_properties(self) -> Dict[str, Any]:
```

File: aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/data_classes/cloudformation_custom_resource_event.py]:
```suggestion
        return self.get("OldResourceProperties") or {}
```

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py:25]: Great catch!

File: aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/cloudformation_custom_resource.py:30]: Great catch!

File: docs/utilities/data_classes.md
Comment (leandrodamascena) [inline docs/utilities/data_classes.md:501]: Please move this example to the `examples/event_sources/src` folder and include it here. This way we can be sure that it will execute linter, typing check and others against this example.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @phipag! Super nice this PR! I just left some few comments to make it even better.

Please let me know if you have any question.

**Comment by phipag** [comment]
Thanks for your review. All makes sense. I’ll send an update on Tuesday. 👍🏻

**Comment by phipag** [comment]
I addressed your comments @leandrodamascena and pushed a new commit.

**Comment by leandrodamascena** [comment]
> I addressed your comments @leandrodamascena and pushed a new commit.

Super thanks @phipag! I see some errors in CI, but those are not related with this PR and I'm fixing this.

I'm going to review this PR today/tomorrow, ok?

**Comment by phipag** [comment]
Sure. No problem. 

Should I merge from develop next time before sending an update?

**Comment by leandrodamascena** [comment]
> Should I merge from develop next time before sending an update?

Hmm, you don't need this unless we see some conflict that blocks the merge. But in this case the problem is because a PR was approved this morning and the linter is complaining. Basically it was my fault because I didn't run the linter locally #mybad 🫨

File: tests/unit/parser/test_cloudformation_custom_resource.py
Comment (leandrodamascena) [inline tests/unit/parser/test_cloudformation_custom_resource.py]:
```suggestion
    raw_event = load_event("cloudformationCustomResourceDelete.json")
```

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hey @phipag! We need to make a very small change before we merge. 🚀

**Comment by phipag** [comment]
Good catch. I'll fix it tomorrow morning! 👍🏻

**Comment by phipag** [comment]
Hey @leandrodamascena I pushed the update.

**Review by leandrodamascena — APPROVED** [review]
Hey @phipag! Thank you for your first contribution to the project. The code is clean and working perfectly for what was proposed. Feel free to contribute as many times as you like! 🚀

**Comment by phipag** [comment]
Awesome! 🙌

**Pull request merged**
